### DISCUSSION TITLE
How to control mutation cache duration

### DISCUSSION THREAD

**Discussion by Noitidart** [TanStack/query#2768] [Q&A] [answered]
It seems all mutations are cached in queryClient.getMutationCache indefinitely. Is it possible to control how long it stays for? I provide a mutationKey.

**Comment by TkDodo** [comment]
I don't think we have a `cacheTime` for mutations

**Reply by allindeveloper** [reply]
Hi @TkDodo 
Can i cache mutations using ReactQuery i've been wanting to do that but it doesn't work.. i just need it to cache the request call and not make the API call again except if the page is refreshed.
Can i achieve this ?

**Reply by Noitidart** [reply]
Oh thanks at @tkDodo sorry I missed that reply. Dang. How long do mutations sit in MutationCache? I am currently doing:

```
const successMutation = queryClient
      .getMutationCache()
      .getAll()
      .find(
        mutation =>
          mutation.state.status === 'success' &&
          mutation.options.mutationKey === mutationKey
      );
```

but not sure if this will stop working after say 5min.

**Reply by TkDodo** [reply]
if I've read the current code correctly, the mutationCache will only be cleared once you call `queryClient.clear()`, which clears both the queryCache and the mutationCache.

**Reply by Noitidart** [reply]
ohhh, thank you! i'm thinking of not using mutations then, it seems a little odd. it can fill up with tons of space i do a lot of mutations. or maybe theres some way to do a mutation with `useQuery`, do you know by chance? I'll also explore. I'm thinking something like this:


```
  const report = useQuery(
    'report',
    (data) => FetchService.post(`api/v1/report`, data),
    {
      retryOnMount: false,
      cacheTime: 60,
      staleTime: 60,
      enabled: false,
      refetchOnMount: false,
      refetchOnWindowFocus: false,
      refetchOnReconnect: false
    }
  );
```

Then I'll do a `report.refetch()`, but do you know how I can modify data payload? Like `report.refetch({ foo: true })`?

**Reply by Noitidart** [reply]
So I was thinking more, and is the mutation cache growing infinitely, and only clears on restart of app? Isn't this a problem? How have others addressed this? Is it possible to clear the cache on unmount of the `useMutation`?

**Reply by TkDodo** [reply]
I think it grows infinitely, yes. Tbh, I’m not sure how this was intended. State of mutations is not shared like state of queries, so I think if you mount the same mutation again, you would still not have the data from the last time, which I think is fine, but it means that the “old” data is indeed just lying around in the cache. Could you validate / verify if that’s the case in a codesandbox?

Personally, i think removing the cache entry after the observer unmounts is a good idea, but it’s also technically a breaking change. Maybe we have to add a cacheTime ?

**Reply by TkDodo** [reply]
A quick user-land solution that comes to mind would be to e.g. check for useIsMutating() and run some cleanup when it goes to zero, like an effect that removes all mutations that have no observers

**Reply by Noitidart** [reply]
Thanks very much for your feedback. I think I can't clear from mutation cache when useIsMutating goes to 0 because maybe some things are still using the mutation.isError .isLoading etc. I would need to clear from cache on unmount.

Way appreciate your chat.

Yep I'll create a codesandbox to verify if mutations sit in cache now. Thanks for the work with me!

**Reply by TkDodo** [reply]
Yes, you’d need to further check for mutations that have no observers. I just thought that whenever a mutation finishes might be a good point in time to cleanup old mutations :)

**Reply by Noitidart** [reply]
Hey hey so I made a sandbox that shows a mutate button, pressing that does a mutation with key "primaryMutation". We see below the mutationCache continues to grow. Interestingly, when I destroy the `useMutation`, the observer isn't removed right away, I'm not sure why. May you please share some advice, I'll keep trying too in meantime - https://codesandbox.io/s/keen-darwin-6n1kt?file=/src/App.js

**Reply by TkDodo** [reply]
> the observer isn't removed right away, I'm not sure why. 

we are not notifying the cache when an observer is added or removed for a mutation - we do that for a query though. if we had that, I think your custom subscription (basically, you just made devtools for mutations 😆 ) would update as well and show zero observers.

question would be: what to do then? Could we just remove mutations from the cache once the observers go to zero? I think that would interfere with the feature that e.g. `onSuccess` callbacks of `.mutate` will always be called, even if you have already unmounted the mutation.

Would the best thing be to add a `cacheTime` to the mutation config that maybe also defaults to 5 minutes?

@tannerlinsley could you help us out here?

**Reply by Noitidart** [reply]
Ah hahaha at mutation devtools! Thanks very much for your collaboration with me!

Just some brainstorm questions:

What is point of mutation cache if when another useMutation mounts, it doesn't reuse that entry in cache? (ideally there is a way to reuse mutation by mutationKey, that was my initial expectation, but this is probably unrelated to the current talk)

If no point to the cache, then onSuccess callback of .mutate can call just fine right, with or without cache?

**Reply by TkDodo** [reply]
> What is point of mutation cache if when another useMutation mounts, it doesn't reuse that entry in cache? (ideally there is a way to reuse mutation by mutationKey, that was my initial expectation, but this is probably unrelated to the current talk)

there is an open issue for that, but i don’t think is that easy: https://github.com/tannerlinsley/react-query/issues/2304

> If no point to the cache, then onSuccess callback of .mutate can call just fine right, with or without cache?

What I wanted to say is that callbacks on useMutation are _always_ executed, even if the last observer is already unmounted. Callbacks on mutate are only executed if the observer is still mounted at the time when the execution should happen. I’ve written about that here: https://tkdodo.eu/blog/mastering-mutations-in-react-query#some-callbacks-might-not-fire

so my fear is that if we garbage collect as soon as the last observer unsubscribes, we would break that feature and potentially callbacks on useMutation would also be skipped..

**Reply by Noitidart** [reply]
Oh darn I see if last observer is gone, callbacks will skip. The only reason I needed useMutation is because it has nice state for loading, error, idle, status. If not for this, then useMutation is nothing special right?

Anyway to `useQuery` but pass in parameters for the query? Like:

```
const foo = useQuery();

foo.refetch({ with these params })
```

Is it possible to remove single entry from MutationCache? I think I only see remove all or none.

**Reply by TkDodo** [reply]
> Anyway to useQuery but pass in parameters for the query? 

You need to use local state and put the params into the query key, but that's a different discussion please.

> Is it possible to remove single entry from MutationCache

`queryClient.getMutationCache().remove(queryClient.getMutationCache().find(...))`

**Reply by Noitidart** [reply]
Thanks so much for answering these while Im mixing the chat up. Really appreciate it!

I'm thinking maybe do that remove part in every `mutate` callback, what do you think? So that allows useMutation and then mutate callback to fire?

**Comment by Noitidart** [comment, marked as answer]
Thank you @TkDodo for this! If anyone comes to this topic there is a mutation cache time feature now! - https://github.com/tannerlinsley/react-query/pull/2963

**Reply by CapitaineToinon** [reply]
For anyone finding this answer today and wondering why the option seems to be gone, it has been renamed to `gcTime` in version 5

https://tanstack.com/query/latest/docs/framework/vue/guides/migrating-to-v5#rename-cachetime-to-gctime
